### PR TITLE
<any>: std::any links when exceptions are disabled

### LINKED ISSUE #990: <any>: std::any doesn't link when exceptions are disabled
**Describe the bug**
`std::any` doesn't link when exceptions are disabled

The issue occurs regardless of `/EH` options; `_HAS_EXCEPTIONS=0` matters here. 
The issue occurs with `/MD` or `/MDd`, not with `/MT` or `/MTd`.

**Command-line test case**
```

D:\Temp2>type repro.cpp
#include <any>

int main() {
    return std::any_cast<int>(std::any{});
}

D:\Temp2>cl /D_HAS_EXCEPTIONS=0 /MD /std:c++17 repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.27.29009.1 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

repro.cpp
Microsoft (R) Incremental Linker Version 14.27.29009.1
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:repro.exe
repro.obj
repro.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __thiscall std::bad_cast::bad_cast(char const *)" (__imp_??0bad_cast@std@@QAE@PBD@Z) referenced in function "public: __thiscall std::bad_any_cast::bad_any_cast(void)" (??0bad_any_cast@std@@QAE@XZ)
repro.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: virtual __thiscall std::bad_cast::~bad_cast(void)" (__imp_??1bad_cast@std@@UAE@XZ) referenced in function "public: virtual __thiscall std::bad_any_cast::~bad_any_cast(void)" (??1bad_any_cast@std@@UAE@XZ)
repro.obj : error LNK2001: unresolved external symbol "protected: virtual void __thiscall std::bad_cast::_Doraise(void)const " (?_Doraise@bad_cast@std@@MBEXXZ)
repro.exe : fatal error LNK1120: 3 unresolved externals
```

**Expected behavior**
The program should link.

**STL version**
```
Microsoft Visual Studio Professional 2019 Preview
Version 16.7.0 Preview 3.1
```

**Additional context**
`_HAS_EXCEPTIONS=0` is unsupported, see https://github.com/microsoft/STL/issues/202#issuecomment-545235685

This item is also tracked on Developer Community as DevCom-376072 and by Microsoft-internal VSO-725649 / [AB#725649](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/725649).

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #990

### CODE REVIEW

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
After looking at this, I agree that the change is correct and ABI-safe.

Usually, anything mentioning `_CRTIMP2_IMPORT` affects ABI. However, your analysis is correct - these are guarded by `#if !_HAS_EXCEPTIONS` and we never build the STL's DLL in that mode (with the exception of `stl/src/raisehan.cpp` but that doesn't drag in `<typeinfo>`), so they never expand to `__declspec(dllexport)`. Thus the macros should be removed, so that they never expand to `__declspec(dllimport)` for user code.

I have verified this by inspecting the DLL's exports (these types don't appear), and by performing an internal build of the STL with

```cpp
#if !_HAS_EXCEPTIONS

#if defined(CRTDLL2) && defined(_CRTBLD)
#error WOOF
#endif

// ...
```

here; it succeeded, proving that these are indeed never exported.

**Comment by StephanTLavavej** [comment]
I'm (speculatively) mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

File: tests/std/tests/GH_000990_any_link_without_exceptions/test.compile.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+#include <any>
+
+using namespace std;
+
+int test() {
+    // GH-990 <any>: std::any doesn't link when exceptions are disabled
+    return any_cast<int>(any{});
+}
+
+int main() {} // COMPILE-ONLY
```
Comment (CaseyCarter) [inline tests/std/tests/GH_000990_any_link_without_exceptions/test.compile.pass.cpp:13]: How does a compile-only test validate linking? Should we not make this a do-nothing-at-runtime test to ensure that linking works?

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for fixing this long-standing bug! :heart_eyes_cat: :lady_beetle: :tada:

**Pull request merged**